src/WebJobs.Script.WebHost/Metrics/FlexConsumptionMetricsPublisher.cs

```diff
@@ -0,0 +1,290 @@
+        internal long FunctionExecutionCount { get; set; }
+
+        internal long FunctionExecutionTimeMS { get; set; }
+
+        internal long ActiveFunctionCount { get; set; }
+
+        internal bool IsAlwaysReady { get; set; }
+
+        internal string MetricsFilePath { get; set; }
+
+        public void Start()
+        {
+            Initialize();
+
+            _logger.LogInformation($"Starting metrics publisher (AlwaysReady={IsAlwaysReady}, MetricsPath='{MetricsFilePath}').");
+
+            _metricsPublisherTimer = new Timer(OnFunctionMetricsPublishTimer, null, _initialPublishDelay, _metricPublishInterval);
+            _started = true;
+        }
+
+        /// <summary>
+        /// Initialize any environmentally derived state after specialization, prior to starting the publisher.
+        /// </summary>
+        internal void Initialize()
+        {
+            _metricPublishInterval = TimeSpan.FromMilliseconds(_optionsMonitor.CurrentValue.MetricsPublishIntervalMS);
+            _initialPublishDelay = TimeSpan.FromMilliseconds(_optionsMonitor.CurrentValue.InitialPublishDelayMS);
+            _intervalStopwatch = ValueStopwatch.StartNew();
+            MetricsFilePath = _optionsMonitor.CurrentValue.MetricsFilePath;
+
+            IsAlwaysReady = _environment.GetEnvironmentVariable(EnvironmentSettingNames.FunctionsAlwaysReadyInstance) == "1";
+        }
+
+        internal async Task OnPublishMetrics()
+        {
+            if (FunctionExecutionCount == 0 && FunctionExecutionTimeMS == 0 && !IsAlwaysReady)
+            {
+                // no activity to report
+                return;
+            }
+
+            // we've been accumulating function activity for the entire period
+            // publish this activity and reset
+            Metrics metrics = null;
+            lock (_lock)
+            {
+                metrics = new Metrics
+                {
+                    TotalTimeMS = (long)_intervalStopwatch.GetElapsedTime().TotalMilliseconds,
+                    ExecutionCount = FunctionExecutionCount,
+                    ExecutionTimeMS = FunctionExecutionTimeMS,
+                    IsAlwaysReady = IsAlwaysReady
+                };
+
+                FunctionExecutionTimeMS = FunctionExecutionCount = 0;
+                _instanceActivityStopwatch = ValueStopwatch.StartNew();
+            }
+
+            await PublishMetricsAsync(metrics);
+        }
+
+        private async void OnFunctionMetricsPublishTimer(object state)
+        {
+            await OnPublishMetrics();
+        }
+
+        private async Task PublishMetricsAsync(Metrics metrics)
+        {
+            try
+            {
+                bool metricsPublishEnabled = !string.IsNullOrEmpty(MetricsFilePath);
+                if (metricsPublishEnabled && !PrepareDirectoryForFile(_fileSystem))
+                {
+                    return;
+                }
+
+                string metricsContent = JsonConvert.SerializeObject(metrics);
```

Comment:
Should we consider switching to the source [gen version of Serialize](https://learn.microsoft.com/en-us/dotnet/standard/serialization/system-text-json/source-generation) from STJ ?

Reply:
Might consider that in the future as an optimization if necessary. I think this standard serialization is fine for now, because this isn't a hotpath - we're logging every 5 seconds or so.